File: src/claw_mem/memory/agnostic.py
"""
AgentAgnosticMemory — cross-agent memory format for claw-mem v4.0.

Provides standardized memory records, format conversion between
local and shared representations, PII filtering, and query filters
for cross-agent memory sharing.
"""

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional
import re
import uuid
import time


@dataclass
class MemoryRecord:
    """Standardized memory record for cross-agent sharing.

    Attributes:
        id: Unique record identifier.
        agent_id: The agent that created the memory.
        memory_type: "episodic" / "semantic" / "procedural" / "shared".
        content: Memory content string.
        tags: Categorization tags.
        timestamp: Unix timestamp when created.
        confidence: Confidence score (0.0–1.0).
        source: "local" or "shared".
    """

    id: str
    agent_id: str
    memory_type: str
    content: str
    tags: List[str]
    timestamp: float
    confidence: float = 1.0
    source: str = "local"


# PII patterns for auto-filtering before cross-agent sharing
_PII_PATTERNS = [
    (re.compile(r"[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}"), "[EMAIL]"),
    (re.compile(r"\b\d{3}[-.]?\d{3}[-.]?\d{4}\b"), "[PHONE]"),
    (re.compile(r"\b\d{3}[-.]?\d{2}[-.]?\d{4}\b"), "[SSN]"),
    (re.compile(r"(?:api[_-]?key|apikey|secret[_-]?key|access[_-]?token)\s*[:=]\s*[\w\-]+", re.IGNORECASE), "[API_KEY]"),
    (re.compile(r"sk-[a-zA-Z0-9]{20,}"), "[API_KEY]"),
]


class AgentAgnosticMemory:
    """Static utilities for cross-agent memory format conversion.

    Example:
        >>> record = AgentAgnosticMemory.to_shared_format(
        ...     {"content": "hello", "tags": ["greeting"]},
        ...     agent_id="agent1",
        ... )
        >>> local = AgentAgnosticMemory.from_shared_format(record)
        >>> filt = AgentAgnosticMemory.create_filter(agent_id="agent1")
    """

    @staticmethod
    def to_shared_format(
        memory: Dict[str, Any], agent_id: str
    ) -> MemoryRecord:
        """Convert a local memory dict to shared MemoryRecord format.

        Args:
            memory: Dictionary with memory data (content, tags, etc.).
            agent_id: ID of the agent that created the memory.

        Returns:
            Standardized MemoryRecord.
        """
        content = str(memory.get("content", ""))
        if not content.strip():
            raise ValueError("Content is required for MemoryRecord")

        # Strip PII before sharing
        clean_content = AgentAgnosticMemory._strip_pii(content)

        return MemoryRecord(
            id=memory.get("id", str(uuid.uuid4())),
            agent_id=agent_id,
            memory_type=memory.get("memory_type", "shared"),
            content=clean_content,
            tags=list(memory.get("tags", [])),
            timestamp=memory.get("timestamp", time.time()),
            confidence=memory.get("confidence", 1.0),
            source="shared",
        )

    @staticmethod
    def from_shared_format(record: MemoryRecord) -> Dict[str, Any]:
        """Convert a shared MemoryRecord back to local format.

        Args:
            record: The shared MemoryRecord.

        Returns:
            Dictionary in local memory format.
        """
        return {
            "id": record.id,
            "content": record.content,
            "memory_type": record.memory_type,
            "tags": record.tags,
            "timestamp": record.timestamp,
            "confidence": record.confidence,
            "agent_id": record.agent_id,
            "source": record.source,
        }

    @staticmethod
    def create_filter(
        agent_id: Optional[str] = None,
        memory_type: Optional[str] = None,
        tags: Optional[List[str]] = None,
        since: Optional[float] = None,
        until: Optional[float] = None,
        min_confidence: float = 0.0,
    ) -> Dict[str, Any]:
        """Create a filter dict for querying MemoryPool.

        Args:
            agent_id: Filter by agent ID.
            memory_type: Filter by memory type.
            tags: Filter by tags (any match).
            since: Filter records after timestamp.
            until: Filter records before timestamp.
            min_confidence: Minimum confidence threshold.

        Returns:
            Filter dictionary for MemoryPool.query().
        """
        filt: Dict[str, Any] = {}
        if agent_id is not None:
            filt["agent_id"] = agent_id
        if memory_type is not None:
            filt["memory_type"] = memory_type
        if tags is not None:
            filt["tags"] = tags
        if since is not None:
            filt["since"] = since
        if until is not None:
            filt["until"] = until
        if min_confidence > 0.0:
            filt["min_confidence"] = min_confidence
        return filt

    @staticmethod
    def _strip_pii(content: str) -> str:
        """Strip known PII patterns from content.

        Args:
            content: Original content string.

        Returns:
            Content with PII patterns replaced by placeholders.
        """
        for pattern, replacement in _PII_PATTERNS:
            content = pattern.sub(replacement, content)
        return content

File: src/memory/agnostic.ts
// Copyright 2026 Peter Cheng
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * AgentAgnosticMemory - cross-agent memory format
 *
 * Provides standardized memory records, format conversion between
 * local and shared representations, PII filtering, and query filters.
 */

import { randomUUID } from "crypto";

// ── PII patterns ──────────────────────────────────────────────────────

const PII_PATTERNS: [RegExp, string][] = [
  [/[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/g, "[EMAIL]"],
  [/\b\d{3}[-.]?\d{3}[-.]?\d{4}\b/g, "[PHONE]"],
  [/\b\d{3}[-.]?\d{2}[-.]?\d{4}\b/g, "[SSN]"],
  [
    /(?:api[_-]?key|apikey|secret[_-]?key|access[_-]?token)\s*[:=]\s*[\w-]+/gi,
    "[API_KEY]",
  ],
  [/sk-[a-zA-Z0-9]{20,}/g, "[API_KEY]"],
];

// ── MemoryRecord ──────────────────────────────────────────────────────

export interface MemoryRecord {
  id: string;
  agent_id: string;
  memory_type: string;
  content: string;
  tags: string[];
  timestamp: number;
  confidence: number;
  source: string;
}

// ── AgentAgnosticMemory ───────────────────────────────────────────────

export class AgentAgnosticMemory {
  /**
   * Convert a local memory dict to shared MemoryRecord format.
   *
   * @param memory - Dictionary with memory data (content, tags, etc.)
   * @param agentId - ID of the agent that created the memory
   * @returns Standardized MemoryRecord
   */
  static to_shared_format(
    memory: Record<string, unknown>,
    agentId: string,
  ): MemoryRecord {
    const content = String(memory.content ?? "");
    if (!content.trim()) {
      throw new Error("Content is required for MemoryRecord");
    }

    const cleanContent = AgentAgnosticMemory._strip_pii(content);

    return {
      id: (memory.id as string) ?? randomUUID(),
      agent_id: agentId,
      memory_type: (memory.memory_type as string) ?? "shared",
      content: cleanContent,
      tags: [...((memory.tags as string[]) ?? [])],
      timestamp: (memory.timestamp as number) ?? Date.now() / 1000,
      confidence: (memory.confidence as number) ?? 1.0,
      source: "shared",
    };
  }

  /**
   * Convert a shared MemoryRecord back to local format.
   *
   * @param record - The shared MemoryRecord
   * @returns Dictionary in local memory format
   */
  static from_shared_format(record: MemoryRecord): Record<string, unknown> {
    return {
      id: record.id,
      content: record.content,
      memory_type: record.memory_type,
      tags: record.tags,
      timestamp: record.timestamp,
      confidence: record.confidence,
      agent_id: record.agent_id,
      source: record.source,
    };
  }

  /**
   * Create a filter dict for querying MemoryPool.
   *
   * @param agentId - Filter by agent ID
   * @param memoryType - Filter by memory type
   * @param tags - Filter by tags (superset match)
   * @param since - Filter records after timestamp
   * @param until - Filter records before timestamp
   * @param minConfidence - Minimum confidence threshold
   * @returns Filter dictionary for MemoryPool.query()
   */
  static create_filter(
    agentId?: string,
    memoryType?: string,
    tags?: string[],
    since?: number,
    until?: number,
    minConfidence: number = 0.0,
  ): Record<string, unknown> {
    const filt: Record<string, unknown> = {};
    if (agentId != null) filt.agent_id = agentId;
    if (memoryType != null) filt.memory_type = memoryType;
    if (tags != null) filt.tags = tags;
    if (since != null) filt.since = since;
    if (until != null) filt.until = until;
    if (minConfidence > 0.0) filt.min_confidence = minConfidence;
    return filt;
  }

  /**
   * Strip known PII patterns from content.
   *
   * @param content - Original content string
   * @returns Content with PII patterns replaced by placeholders
   */
  static _strip_pii(content: string): string {
    for (const [pattern, replacement] of PII_PATTERNS) {
      content = content.replace(pattern, replacement);
    }
    return content;
  }
}
